### PR TITLE
Implement VK_EXT_debug_utils 

### PR DESCRIPTION
**Pull request by jonathansty**
* [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
* [x] Updated documentation to reflect any user-facing changes - in this repository
* [x] Update example to use debug utils and demonstrate common use cases

Implements the `VK_EXT_debug_utils` extensions into vulkano. This allows you to setup debug callbacks, filters, tag objects and use debug labels for programs such as renderdoc.

Currently the functionality to set labels on queues is not exposed/implemented yet. Although I would want to leave this for a future issue as I'm not sure what part should 'own' these functions.

### CODE REVIEW

**Comment by jonathansty** [comment]
#1051 #1066

**Comment by nicokoch** [comment]
Note that this Extension is deprecated and should probably be implemented using vk_ext_debug_utils

**Comment by jonathansty** [comment]
> vk_ext_debug_utils

Ah thanks! I just used the extensions that were already present in the code but I suppose I could easily change this to use the newer extension. I'll have a look in to it.

**Comment by rukai** [comment]
Sure, here are some initial thoughts:
*   I think it should return an Err rather then failing silently.
*   I rearranged extensions and features last night. So although your current code isn't affected by it. You might want to `git fetch origin; git rebase origin/master` to avoid merge conflicts in the future.
*   I think we should remove `VK_EXT_debug_marker` and `VK_EXT_debug_report` now that they are deprecated.
*  Some sort of example would be good, maybe extend the `debug.rs` example or create a new example based off `basic-compute-shader.rs` with the comments stripped out.

**Comment by rukai** [comment]
Yeah wow, thanks for demonstrating the issue.
At first I thought just unwraping the value would be fine, but I guess this debug stuff is meant to be left in the application, which would lead to it randomly panicking on some platforms.

So I think silently ignoring is the best solution.
However you should clearly document that is the case and why. Maybe something like:
If the `ext_debug_utils` `DeviceExtension` is not available this method will fail silently. If this method appears to have no effect, verify that the device extension is available on your platform.

**Comment by jonathansty** [comment]
@rukai  I'm at a point now where I'm happy with the current changes. 

The only thing that I haven't done yet is implemented ways to access `QueueBeginDebugUtilsLabelExt`, `QueueEndDebugUtilsLabelEXT` and `QueueInsertDebugUtilsLabelEXT` because I'm not entirely sure where these should live. 

I'm also unsure whether or not they should be under the device pointers in this case.

**Comment by nicokoch** [comment]
The DebugMessenger should probably live inside `Instance`. Same reason as stated in https://github.com/vulkano-rs/vulkano/issues/821

**Comment by jonathansty** [comment]
@nicokoch sure that makes sense although I want to limit the scope of this PR seeing as it's already growing quite a lot. 

I tried adding it quick and dirty to the instance just for testing but was met with problems such as the `instance::new` function returning an `Arc` and I don't have enough rust knowledge to actually know how to nicely solves these issues.

**Comment by rukai** [comment]
```
warning: unused import: `c_char`
  --> vulkano/src/instance/debug.rs:44:20
   |
44 | use std::os::raw::{c_char, c_void};
   |                    ^^^^^^
   |
```

File: examples/src/bin/debug.rs [hunk 1]
```diff
@@ -27,8 +41,9 @@ fn main() {
     //
     // .. but if you just want a template of code that has everything ready to go then follow
     // this example. First, enable debugging using this extension: VK_EXT_debug_report
+    // For extra debugging features also enable: VK_EXT_debug_utils
```
Comment (rukai) [inline examples/src/bin/debug.rs]: Combine with the previous sentence and reword. VK_EXT_debug_report shouldn't be mentioned, its deprecated.

File: examples/src/bin/debug.rs [hunk 2]
```diff
@@ -103,5 +136,59 @@ fn main() {
...
+    let buffer1 = CpuAccessibleBuffer::from_iter(device.clone(), BufferUsage::all(), (0..512*512).map(|e| e)).unwrap();
+    let img = StorageImage::new(device.clone(), vulkano::image::Dimensions::Dim2d{width:512,height:512}, Format::R8Uint, Some(queue.family())).unwrap();
+
+    // VK_EXT_debug_utils allows setting object tags which will be reported whenever a message that is relevant to these objects
+    // is being reported. See the debug callback for information on how to access them.
+    device.set_object_name(img.inner().image, &CString::new("DestinationImage").unwrap()).ok();
+    device.set_object_name(buffer1.inner().buffer, &CString::new("DestinationImage").unwrap()).ok();
+
+    // Instead of using something like an AutoCommandBufferBuilder this is using a UnsafeCommandBufferBuilder so we can
+    // forcefully insert incorrect function calls and trigger validation errors and a crash. In a real application you should probably.
```
Comment (rukai) [inline examples/src/bin/debug.rs]: probably what?

**Comment by jonathansty** [comment]
@rukai  I've been trying to find another way to reproduce a validation warning/message eg with `UnsafeImage` but I can't seem to get `copy_buffer_to_image` working because we have no access whatsever to any of the unsafe sides of the vulkan API, which imho is a massive roadblock.

Half tempted to just remove the unsafe block of code as it's quite obvious that the seg fault is probably happening because I'm instructing a draw command without any buffers/ pipelines bound.

**Comment by jonathansty** [comment]
@rukai  I've removed the code that caused the segmentation fault and have replaced it with manual `submit_debug_message` calls to trigger some dummy warning and error messages. 

Right now there should also be 1 actual error message about a StorageImage not being in the correct layout when trying to copy a buffer to it.

Any remarks?

**Comment by rukai** [comment]
I'm still getting a segfault, for `cargo run --release --bin debug`
```
List of Vulkan debugging layers available to use:
	VK_LAYER_LUNARG_screenshot
	VK_LAYER_LUNARG_vktrace
	VK_LAYER_LUNARG_api_dump
	VK_LAYER_LUNARG_device_simulation
	VK_LAYER_LUNARG_monitor
	VK_LAYER_GOOGLE_threading
	VK_LAYER_LUNARG_object_tracker
	VK_LAYER_LUNARG_parameter_validation
	VK_LAYER_GOOGLE_unique_objects
	VK_LAYER_LUNARG_core_validation
	VK_LAYER_LUNARG_standard_validation
	VK_LAYER_VALVE_steam_overlay_32
	VK_LAYER_VALVE_steam_overlay_64
error: This is a debug error message!
Queue Labels(0):
Command Labels(0):
Objects(1):
  Obj: DebugMsg, 0x0, 100
warning: This is a debug warning message!
Queue Labels(0):
Command Labels(0):
Objects(1):
  Obj: DebugMsg, 0x0, 100
Segmentation fault (core dumped)
```

with `cargo run --bin debug` I get a panic :D

```
List of Vulkan debugging layers available to use:
	VK_LAYER_LUNARG_screenshot
	VK_LAYER_LUNARG_vktrace
	VK_LAYER_LUNARG_api_dump
	VK_LAYER_LUNARG_device_simulation
	VK_LAYER_LUNARG_monitor
	VK_LAYER_GOOGLE_threading
	VK_LAYER_LUNARG_object_tracker
	VK_LAYER_LUNARG_parameter_validation
	VK_LAYER_GOOGLE_unique_objects
	VK_LAYER_LUNARG_core_validation
	VK_LAYER_LUNARG_standard_validation
	VK_LAYER_VALVE_steam_overlay_32
	VK_LAYER_VALVE_steam_overlay_64
error: This is a debug error message!
Queue Labels(0):
Command Labels(0):
Objects(1):
  Obj: DebugMsg, 0x0, 100
warning: This is a debug warning message!
Queue Labels(0):
Command Labels(0):
Objects(1):
  Obj: DebugMsg, 0x0, 100
thread 'main' panicked at 'function pointer `CmdBeginDebugUtilsLabelEXT` not loaded', vk-sys/src/lib.rs:2911:1
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:227
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:475
   5: std::panicking::continue_panic_fmt
             at libstd/panicking.rs:390
   6: std::panicking::begin_panic_fmt
             at libstd/panicking.rs:345
   7: vk_sys::DevicePointers::load::CmdBeginDebugUtilsLabelEXT
             at vk-sys/src/lib.rs:2822
   8: vk_sys::DevicePointers::CmdBeginDebugUtilsLabelEXT
             at /home/rubic/Projects/Crates/vulkano/vulkano/vk-sys/src/lib.rs:2835
   9: <vulkano::command_buffer::sys::UnsafeCommandBufferBuilder<P>>::begin_debug_label
             at /home/rubic/Projects/Crates/vulkano/vulkano/vulkano/src/command_buffer/sys.rs:1486
  10: debug::main
             at examples/src/bin/debug.rs:221
  11: std::rt::lang_start::{{closure}}
             at /checkout/src/libstd/rt.rs:74
  12: std::panicking::try::do_call
             at libstd/rt.rs:59
             at libstd/panicking.rs:310
  13: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:105
  14: std::rt::lang_start_internal
             at libstd/panicking.rs:289
             at libstd/panic.rs:392
             at libstd/rt.rs:58
  15: std::rt::lang_start
             at /checkout/src/libstd/rt.rs:74
  16: main
  17: __libc_start_main
  18: _start
```

**Comment by jonathansty** [comment]
@rukai  interestingly not getting the seg fault locally although I still do have an unsafe command buffer block being used. I also wonder if there is a problem with the image format `R8Uint` I was converting the unsafe command buffer builder to be safe and it wasn't allowing me to copy from my buffer to an image of format `R8Uint`.

The other message sounds like something is wrong with the SDK. I'm also not able to reproduce this. Is there ways of checking if function pointers are correctly loaded?

**Pull request closed without merging**